### DISCUSSION TITLE
Feature Request: support impersonation to access multiple mailboxes from the same user account

### DISCUSSION THREAD

**Discussion by acran** [Mailu/Mailu#3106] [Feature requests / Ideas]
# Description

Accessing different mailboxes - especially via the webmailer - is currently a bit cumbersome since it requires completely logging out of mailu and logging back in again with another user account.

Instead there should be a simple way, e.g., an additional button in the user list - as a global admin or domain manager - to "_open mailbox_" of another user directly in the webmailer.

# Motivation

Mailu being a perfectly suited tool for self-hosting your e-mails it is likely also used much for hosting personal e-mails and domains. Given the ease of creating new e-mail addresses it is also likely that one user has several different e-mail addresses. Using aliases these can then all be delivered to the same mailbox.
But in some cases it may make sense - or be preferred by the user - to also use separate mailboxes for the addresses. But accessing those mailboxes then also requires separate logins.

## Further Motivation

Using an [external authentication](https://mailu.io/2.0/configuration.html#header-authentication-using-an-external-proxy) (exclusively) in such a case adds additional hurdles to this, because it then requires to also create and grant access to new user accounts in the external identity provider just to be able to login to mailu - although it is still only the same "user" accessing those mailboxes.

# Possible solutions

This can be addressed in multiple stages - while all of the below should also only apply to users who could accessing the target mailboxes anyway by, e.g., resetting the account's password, i.e., global administrators or domain managers:

1. allow admins to manage auth-tokens of their users
In the case of using an external authentication this would at least allow the admin to create a new auth-token in order to access the mailbox using any mail client (other than the webmailer).
The complexity of implementing this should be quite low and mostly only require adapting the admin view.

2. allow admins to log into the webmailer as another user
This would give an easy and convenient direct access to the mailbox. Implementing this should also be possible with few less invasive changes which should mostly be limited to `verify_temp_token`:

https://github.com/Mailu/Mailu/blob/df5b48a52dc78db653d2e7dd4bc5036d179b73e0/core/admin/mailu/utils.py#L502-L511

3. implement a full impersonation feature
This would allow the admin to access the whole mailu backend as the target user. If already having implemented 1. & 2. and given the existing access of admins this would the be a bit redundant probably. Also implementing this _may_ require some more major refactoring? On the other hand it would make implementing 1./2. obsolete.

I could maybe have a try on implementing the chosen solution myself and submitting a PR given a green light by the mailu maintainers.

# Possible future expansions and related issues

Thinking further I'd like to also throw in the following ideas - **but purely for discussion and way out of scope for this feature request**:

1. fine grained mailbox access and delegation
Adding a finer grained access model to grant access to a user's mailbox to other users - without them being domain admins. This would enable to have a mailbox used as a kind of "shared mailbox" between multiple (non-admin) users. Also it would allow users to delegate access to their mailboxes, e.g., for a holiday substitution. Ideally for that users could manage the access to their mailbox themselves.

2. splitting the concepts of "users" and "mailboxes"
Currently each user corresponds to a mailbox and vice versa. Splitting these to concepts would allow to create mailboxes without having to create a user with its own login credentials. On the other hand a user could be assigned multiple separate mailboxes - or even no mailboxes at all for users that should be send-only, see #2846

**Comment by nextgens** [comment]
(1) is already possible; Any admin can edit tokens/application-passwords of their users using the existing interface

**Reply by acran** [reply]
🙈 then I seem to just keep missing it, can't find it even now either. Where exactly can an admin so that?

**Reply by nextgens** [reply]
The link points to ``/admin/token/list/user-to-edit@example.com``

**Reply by acran** [reply]
Ah I see, the URL works fine. But as far as I can see it is not linked anywhere and thus can not conveniently be found.

I'd suggest to add the link to the user list:

~~~diff
diff --git a/core/admin/mailu/ui/templates/user/list.html b/core/admin/mailu/ui/templates/user/list.html
index 45a9cb10..a96fe1df 100644
--- a/core/admin/mailu/ui/templates/user/list.html
+++ b/core/admin/mailu/ui/templates/user/list.html
@@ -34,6 +34,7 @@
     </td>
     <td>
       <a href="{{ url_for('.user_settings', user_email=user.email) }}" title="{% trans %}Settings{% endtrans %}"><i class="fa fa-wrench"></i></a>&nbsp;
+      <a href="{{ url_for('.token_list', user_email=user.email) }}" title="{% trans %}Authentication tokens{% endtrans %}"><i class="fa fa-ticket-alt"></i></a>&nbsp;
       <a href="{{ url_for('.user_reply', user_email=user.email) }}" title="{% trans %}Auto-reply{% endtrans %}"><i class="fa fa-plane"></i></a>&nbsp;
       {%- if config["FETCHMAIL_ENABLED"] -%}
       <a href="{{ url_for('.fetch_list', user_email=user.email) }}" title="{% trans %}Fetched accounts{% endtrans %}"><i class="fa fa-download"></i></a>&nbsp;
~~~

Should I open a PR with this?

**Reply by nextgens** [reply]
Yes.

Implementing (2) is slightly trickier than what you have suggested; if you include it on the PR I will review it.

**Reply by acran** [reply]
I submitted this in 88c972afe34523d276be31be1176c3154bada8b0 with #3163

**Comment by acran** [comment]
With 1. already implemented, only 2. would be needed to realize a more or less full impersonation feature.

So what are the opinions on implementing 2.?

**Reply by nextgens** [reply]
I think that we should display a non dismissible "alert" when we are impersonating ... that most likely means framing the webmail.

As for how we do it technically, it's not hard but I think it does involve changing PHP stubs of both webmails.

**Reply by acran** [reply]
> Implementing (2) is slightly trickier than what you have suggested; if you include it on the PR I will review it.

yes, changing `verify_temp_token` to accept tokens for _delegated_ mailbox access for admins should be quite simple. The "hard" part I guess is how to best pass the currently open mailbox to the webmail and how to reset it when done impersonating.

My approach would be to store the currently active mailbox as a new field in the session which is then used by `verify_temp_token`  and `gen_temp_token`.
To select and store the selected mailbox in the session I would add a new route - e.g., `/user/webmail/<path:user_email>` - which just stores the `user_email` in the current session and then redirects to the webmail.

To "reset"/disable impersonation and return to their own mailbox the webmailer link in the sidebar navigation is then replaced by `/user/webmail/<admin_email>` .

Pro: this would work without any changes in the webmailers
Con: it might be a bit non-obvious to the users that they're (still) logged into another mailbox and how to get back to their own.

> I think that we should display a non dismissible "alert" when we are impersonating ...

That's probably a good idea.

For this we could add the e-mail of the impersonating user in an additional header to the webmail, e.g., `X-Original-User`:

https://github.com/Mailu/Mailu/blob/df5b48a52dc78db653d2e7dd4bc5036d179b73e0/core/admin/mailu/internal/views/auth.py#L82-L85

Then the webmail can check if `X-Original-User` is present and different from `X-User` and if so inject an alert.

---

I'm not 100% sure about `/user/webmail/<path:user_email>` + just storing the impersonated e-mail in the session being the cleanest approach here.
But if acceptable, I could try to implement it as described above^^

**Reply by acran** [reply]
I now created PR #3163 for an implementation of the above.

It adds a simple button to the user list to open the webmailer as another user. For this the button navigates to the new route `/user/webmail/<path:user_email>` which just stores the specified user as current impersonated user in the session and then redirects to the webmail `sso.php` for login.

The webmailer receives the impersonated username in `X-Remote-User` with the token in `X-Remote-User-Token` - as before such that webmailers shouldn't really need any changes to work with this.

Additionally `X-Remote-Original-User` now contains the actual/originally logged in user, so the webmail can display an alert if this differs from the mailbox user - which I implemented for roundcube.
